docs/freqai-reinforcement-learning.md

```diff
@@ -0,0 +1,230 @@
+
+    def populate_exit_trend(self, df: DataFrame, metadata: dict) -> DataFrame:
+        exit_long_conditions = [df["do_predict"] == 1, df["&-action"] == 2]
+        if exit_long_conditions:
+            df.loc[reduce(lambda x, y: x & y, exit_long_conditions), "exit_long"] = 1
+
+        exit_short_conditions = [df["do_predict"] == 1, df["&-action"] == 4]
+        if exit_short_conditions:
+            df.loc[reduce(lambda x, y: x & y, exit_short_conditions), "exit_short"] = 1
+
+        return df
+```
+
+It is important to consider that `&-action` depends on which environment they choose to use. The example above shows 5 actions, where 0 is neutral, 1 is enter long, 2 is exit long, 3 is enter short and 4 is exit short. 
+
+## Configuring the Reinforcement Learner
+
+In order to configure the `Reinforcement Learner` the following dictionary must exist in the `freqai` config:
+
+```json
+        "rl_config": {
+            "train_cycles": 25,
+            "add_state_info": true,
+            "max_trade_duration_candles": 300,
+            "max_training_drawdown_pct": 0.02,
+            "cpu_count": 8,
+            "model_type": "PPO",
+            "policy_type": "MlpPolicy",
+            "model_reward_parameters": {
+                "rr": 1,
+                "profit_aim": 0.025
+            }
+        }
+```
+
+Parameter details can be found [here](freqai-parameter-table.md), but in general the `train_cycles` decides how many times the agent should cycle through the candle data in its artificial environment to train weights in the model. `model_type` is a string which selects one of the available models in [stable_baselines](https://stable-baselines3.readthedocs.io/en/master/)(external link).
+
+!!! Note
+    Remember that the general `model_training_parameters` dictionary should contain all the model hyperparameter customizations for the particular `model_type`. For example, `PPO` parameters can be found [here](https://stable-baselines3.readthedocs.io/en/master/modules/ppo.html).
+
+## Creating the reward
+
+As you begin to modify the strategy and the prediction model, you will quickly realize some important differences between the Reinforcement Learner and the Regressors/Classifiers. Firstly, the strategy does not set a target value (no labels!). Instead, you set the `calculate_reward()` function inside the `ReinforcementLearner.py` file. A default `calculate_reward()` is provided inside `prediction_models/ReinforcementLearner.py` to demonstrate the necessary building blocks for creating rewards. It is inside the `calculate_reward()` where creative theories about the market can be expressed. For example, you can reward your agent when it makes a winning trade, and penalize the agent when it makes a losing trade. Or perhaps, you wish to reward the agent for entering trades, and penalize the agent for sitting in trades too long. Below we show examples of how these rewards are all calculated:
+
+```python
+    class MyRLEnv(Base5ActionRLEnv):
```

Comment:
> I do guess that you did an evaluation on this and somehow got to the conclusion that this is the right architecture ?

I'd rather not make users handle yet a *fourth* file (strat + PredictionModel + config + your suggested environment file). Best to keep the same number of files that FreqAI already requires for customization, strat + PredictionModel + config (already too many in my opinion). This also means we reduce the number of flags required for users to learn and carry around on their launch commands (which reduces points of error). 

With the present structure, the user only needs their custom `ReinforcementLearner_custom.py` with the `calculate_reward()` defined inside (or any other function they want to override from the inheritance structure). Especially considering the majority of users will only want to modify the `calculate_reward()`, this makes much more sense than creating and handling a fourth file. 

If we want to support power users, we can add an optional environment resolution in the future (like what we do with the prediction model). However, I'd prefer to release and see the need organically requested, instead of just make a bunch of machinery that gets neglected in preference of defining the single `calculate_reward()` function. 

> the same subclass name, unless he aims to change /overwrite all occurances of MyRLEnv (it's not 100% clear that the name of this subclass MUST be MyRLEnv, and can't be MyAwesomeRLEnv

Yes `MyRLEnv` needs to stay the same. It is yet another function to override. Just as all users need to stick to `custom_exit()` for naming their `custom_exit()`. All users need to stick to `MyRLEnv` to create their custom environment.

> which is wrong, a user needs to:

    create a new model in user_data/freqaimodels - which subclasses ReinforcementLearner (or the multiprocessing counterpart). This class doesn't need to contain anything other than the below MyRLEnv class.
    
While what you say is correct, and works quite effectively for a user who doesnt wish to see the `def fit()` function in their custom file, I created the FreqAI structure with the idea that the lowest level of the inheritance tree (the user defined level), inherits from a `Base*Model`. E.g. `CatboostRegressor(BaseRegressionModel)` and in the present case `ReinforcmentLearner(BaseReinforcementLearnerModel)`. That means the user is forced to engage at that level, and gives them some basic understanding "oh this is the `def fit()` function, why and what am I fitting?" However, we can certainly follow your recommendation and tell users to inherit from `ReinforcmentLearner` if they want to define a simple `calculate_reward()` function. 

>From reading the docs alone, my initial attempt was wrong / not working - which had me do the remaining review to figure out why it didn't work - which is a sign that the docs around this can be improved

I am improving the docs to ensure users have a better understanding of how to define a custom `calculate_reward()` without dealing with the `def fit()`.

Reply:
I have added quite a bit more explanation in the documentation about the easiest way to create a `custom_reward()` [here](https://github.com/freqtrade/freqtrade/blob/9f13d99b999047237055d0650812e58fef127ab5/docs/freqai-reinforcement-learning.md#creating-a-custom-reward-function).

Reply:
> I'd rather not make users handle yet a fourth file 

As mentioned - pretty good thought about it, and this is also what i assumed is the reason :+1: 
 it took me like 2 days to write this comment alone (as you'll have noticed, it contains parts from all over) - and i rewrote most sections 2-3 times as i discovered new things :laughing: 


---

The idea is one thing (everyone can override everything) - but as you correctly point out above - the majority of users won't use it that way (override `fit()` for example). if they do, we assume them as pretty advanced - so they will probably figure it out anyway.